src/cls/IPM/ResourceProcessor/WebApplication.cls

```diff
@@ -0,0 +1,320 @@
+     * attribute as the attribute list is just all non-private/transient properties 
+     * in Security.Applications.
+     */
+    Set key = ""
+    While 1 {
+        Set value = ..ResourceReference.Attributes.GetNext(.key)
+        If (key = "") {
+            Quit
+        }
+        Set eval = ..%Evaluate(value)
+        Set ..SecurityProperties(key) = eval
+    }
+}
+
+/// Called as phase <var>pPhase</var> is executed for the resource. If <var>pResourceHandled</var> is set to true,
+/// then the default behavior for that resource will be bypassed in the current phase.
+Method OnPhase(pPhase As %String, ByRef pParams, Output pResourceHandled As %Boolean = 0) As %Status
+{
+    Set sc = $$$OK
+    Try {
+        If (pPhase = "Validate") {
+            If (..Name = "") {
+                Return $$$ERROR($$$GeneralError,"Name is required for web applications")
+            }
+            Set keyList = ""
+            Set key = ""
+            While 1 {
+                Set value = ..ResourceReference.Attributes.GetNext(.key)
+                If (key = "") {
+                    Quit
+                }
+                Set keyList = keyList _ $ListBuild(key)
+            }
+            Set badKeyList = ..DoPropertiesExist(keyList)
+            If ($ListLength(badKeyList) > 0) {
+                Set error = "The following invalid attributes were found for resource "_..ResourceReference.Name_": "_$ListToString(badKeyList)
+                $$$ThrowStatus($$$ERROR($$$GeneralError,error))
+            }
+        }
+    } Catch (ex) {
+        Set sc = ex.AsStatus()
+    }
+    Return sc
+}
+
+Method OnBeforePhase(pPhase As %String, ByRef pParams) As %Status
+{
+    /*
+     * NOTE: Web app creation is done in OnBeforePhase so it happens before other
+     * resource processors such as FileCopy run.
+     */
+    Set sc = $$$OK
+    Try {
+        $$$ThrowOnError(##super(pPhase,.pParams))
+        
+        Set tVerbose = $Get(pParams("Verbose"))
+        
+        If ((pPhase = "Configure") || (pPhase = "Activate")) {
+            // Create web application
+            Do ..CreateOrUpdateWebApp(tVerbose)
+        } ElseIf ((pPhase = "Unconfigure") || (pPhase = "Clean")) {
+            // Remove web application
+            Do ..DeleteWebApp(tVerbose)
+            Do ..DeleteCodeReadRole($Get(..SecurityProperties("NameSpace")), tVerbose)
+        }
+    } Catch e {
+        Set sc = e.AsStatus()
+    }
+    Quit sc
+}
+
+Method CreateOrUpdateWebApp(pVerbose As %Boolean = 0) [ Internal ]
+{
+    New $namespace
+    Set $namespace = "%SYS"
+
+    Merge properties = ..SecurityProperties
+
+    // Replace ${codeReadRole} with a role (potentially to be created) with read permission to routine database
+    If ($Data(properties("NameSpace"), ns) # 2) && ($Data(properties("MatchRoles"), mr) # 2) {
+        Set properties("MatchRoles") = ..ReplaceCodeReadRole(mr, ns, pVerbose)
```

Comment:
@isc-kiyer 
I was thinking this isn't a simple string substitution because of the side-effect of creating a role. We also need to remove the role later if the package is uninstalled.